Document about wallet backup and restoration (#22523)
bitcoin/bitcoin: doc/managing-wallets.md

# Managing the Wallet

## 1. Backing Up and Restoring The Wallet

### 1.1 Creating the Wallet

Since version 0.21, Bitcoin Core no longer has a default wallet.
Wallets can be created with the `createwallet` RPC. More information about this command may be found by running `bitcoin-cli help createwallet`.

The following command, for example, creates a descriptor wallet:

`$ bitcoin-cli -named createwallet wallet_name="wallet-01" descriptors=true`

> In the GUI, the `Create a new wallet` button is displayed on the main screen when there is no wallet loaded. Alternatively, there is the option `File` ->`Create wallet`.

Reviewer: Nit: GUI instructions first would be better

Author: Done.

The `descriptors` parameter can be omitted if the intention is to create a legacy wallet.

> 

Reviewer: Nit: Drop blank line, these two sentences are too closely related.

Author: Done.

> For now, the default type is the legacy wallet, but that should change in the near future.

Reviewer: nit

```suggestion
For now, the default type is the legacy wallet, but that is expected to change in a future release.
```

Author: Done.

By default, wallets are created in the `wallets` folder of the data directory, which varies by operating system, as shown below. The user can change the default by using the `-datadir` or `-walletdir` initialization parameters.

| Operating System | Default wallet directory                                    |
| -----------------|:------------------------------------------------------------|
| Linux            | `/home/<user>/.bitcoin/wallets`                             |
| Windows          | `C:\Users\<user>\AppData\Roaming\Bitcoin\wallets`           |
| macOS            | `/Users/<user>/Library/Application Support/Bitcoin/wallets` |

### 1.2 Encrypting the Wallet

The `wallet.dat` file is not encrypted by default and is, therefore, vulnerable if an attacker gains access to the device where the wallet or the backups are stored.

> One way to mitigate this problem is to encrypt the wallet. However, this solution significantly increases the risk of losing coins due to forgotten passphrases. There is no way to recover a passphrase. This tradeoff should be well thought out by the user.

Reviewer: This isn't really solved by wallet encryption. If the attacker has access to the device, they can probably install a keylogger as well.

Wallet encryption really only handles the "little brother" threat.

Reviewer 2: Agreed, tried to point that out [here](https://github.com/bitcoin/bitcoin/pull/22523#discussion_r677926273)

Author: Added.

`$ bitcoin-cli -rpcwallet="wallet-01" walletpassphrase "passphrase" 120`

In the GUI, there is no specific menu item to unlock the wallet. When the user sends bitcoins, the passphrase will be prompted automatically.

### 1.4 Backing Up the Wallet

To backup the wallet, the `backupwallet` RPC or the `Backup Wallet` GUI menu item must be used to ensure the file is in a safe state when the copy is made.

In the RPC, the destination parameter must include the name of the file. Otherwise, the command will return an error message like "Error: Wallet backup failed!" for descriptor wallets. If it is a legacy wallet, it will be copied and a file will be created with the default file name `wallet.dat`.

`$ bitcoin-cli -rpcwallet="wallet-01" backupwallet /home/node01/Backups/backup-01.dat`

In the GUI, the wallet is selected in the `Wallet` drop-down list in the upper right corner. If this list is not present, the wallet can be loaded in `File` ->`Open wallet` if necessary. Then, the backup can be done in `File` -> `Backup Wallet...`.

> This backup file can be stored on one or multiple offline devices, which must be reliable enough to work in an emergency, never be connected to online devices, and be malware free. Backup files can be regularly tested to avoid problems in the future.

Reviewer: "never be connected to online devices" sounds a bit strong. Maybe just express that getting the file can compromise the wallet?

Reviewer 2: Maybe something like: "retrieving the backup file can compromise the wallet if your computer has malware.  Ways to minimize this is to not connect your backup to an online device"?

Author: Added this suggestion.

> If the wallet and backup are lost for any reason, the bitcoins related to this wallet will become permanently inaccessible.

Reviewer: nit

```suggestion
If both the wallet and all backups are lost for any reason, the bitcoins related to this wallet will become permanently inaccessible.
```

Author: Done.

# Managing the Wallet

## 1. Backing Up and Restoring The Wallet

### 1.1 Creating the Wallet

Since version 0.21, Bitcoin Core no longer has a default wallet.

> Wallets can be created with the `createwallet` RPC. More information about this command may be found by running `bitcoin-cli help createwallet`.

Reviewer 2: Include GUI option here, as mentioned by @achow101 ?

Author: Done.

The wallet's private key can also be encrypted in the `createwallet` command via the `passphrase` argument:

`$ bitcoin-cli -named createwallet wallet_name="wallet-01" descriptors=true passphrase="passphrase"`

Note that if the passphrase is lost, all the coins in the wallet will also be lost forever.

### 1.3 Backing Up the Wallet

To backup the wallet, the `backupwallet` RPC or the `Backup Wallet` GUI menu item must be used to ensure the file is in a safe state when the copy is made.

In the RPC, the destination parameter must include the name of the file. Otherwise, the command will return an error message like "Error: Wallet backup failed!" for descriptor wallets. If it is a legacy wallet, it will be copied and a file will be created with the default file name `wallet.dat`.

`$ bitcoin-cli -rpcwallet="wallet-01" backupwallet /home/node01/Backups/backup-01.dat`

In the GUI, the wallet is selected in the `Wallet` drop-down list in the upper right corner. If this list is not present, the wallet can be loaded in `File` ->`Open wallet` if necessary. Then, the backup can be done in `File` -> `Backup Wallet...`.

> This backup file can be stored on one or multiple offline devices, which must be reliable enough to work in an emergency, never have connected to online devices, and be malware free. Backup files can be regularly tested to avoid problems in the future.

Reviewer 2: ```suggestion
This backup file can be stored on one or multiple offline devices, which must be reliable enough to work in an emergency, never be connected to online devices, and be malware free. Backup files can be regularly tested to avoid problems in the future.
```

Author: Done.

Bitcoin Core [version 0.13](https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.13.0.md) introduced HD wallets with deterministic key derivation. With HD wallets, users no longer lose funds when restoring old backups because all addresses are derived from the HD wallet seed.

This means that a single backup is enough to recover the coins at any time. It is still recommended to make regular backups (once a week) or after a significant number of new transactions to maintain the metadata, such as labels. Metadata cannot be retrieved from a blockchain rescan, so if the backup is too old, the metadata will be lost forever.

Wallets created before version 0.13 are not HD and must be backed up every 100 keys used since the previous backup, or even more often to maintain the metadata.

### 1.5 Restoring the Wallet From a Backup

To restore a wallet, first a directory must be created in the `walletdir` with a name that the user wants the wallet to be named, and then copy the backup file into that directory. The copied file must be named `wallet.dat`.

> Then the user can load the wallet passing the name of the  newly created directory as parameter.

Reviewer 2: ```suggestion
Then the user can load the wallet passing the name of the newly created directory as parameter.
```
nit: extra space

Author: Done.

The following command, for example, creates a descriptor wallet:

`$ bitcoin-cli -named createwallet wallet_name="wallet-01" descriptors=true`

In the GUI, the `Create a new wallet` button is displayed on the main screen when there is no wallet loaded. Alternatively, there is the option `File` ->`Create wallet`.

The `descriptors` parameter can be omitted if the intention is to create a legacy wallet.

For now, the default type is the legacy wallet, but that should change in the near future.

By default, wallets are created in the `wallets` folder of the data directory, which varies by operating system, as shown below. The user can change the default by using the `-datadir` or `-walletdir` initialization parameters.

| Operating System | Default wallet directory                                    |
| -----------------|:------------------------------------------------------------|
| Linux            | `/home/<user>/.bitcoin/wallets`                             |
| Windows          | `C:\Users\<user>\AppData\Roaming\Bitcoin\wallets`           |
| macOS            | `/Users/<user>/Library/Application Support/Bitcoin/wallets` |

> ### 1.2 Encrypting the Wallet

Reviewer 3: Perhaps a section on unlocking the wallet? There's the `walletpassphrase` RPC, and the GUI will prompt to unlock when needed.

Reviewer 2: Will re-ACK after this!

Author: Section added.
